src/ranged.cpp

```diff
@@ -1529,6 +1533,31 @@ std::vector<tripoint> target_handler::target_ui( player &pc, target_mode mode,
             pc.last_target_pos = cata::nullopt;
             target = -1;
             break;
+        } else if( action == "zoom_in" ) {
+            g->zoom_in();
+        } else if( action == "zoom_out" ) {
+            g->zoom_out();
+        } else if( action == "MOUSE_MOVE" ) {
+            int max_consume = 10;
+            do {
+                // Below we implement mouse panning. In order to make
+                // it less jerky we rate limit it by only allowing a
+                // panning move during the first iteration of this
+                // mouse move event consumption loop.
+                if( max_consume == 10 ) {
+                    const tripoint edge_scroll = g->mouse_edge_scrolling_terrain( ctxt );
```

Comment:
Looks suspiciously similar to the chunk above, with these couple of lines being a sole difference. Can they be unified?

Reply:
Yes, I noted this in the "Describe the solution" section.

My immediate idea is to refactor the whole loop into a function that looks something like this (I just looked up the syntax for the first time for C++ so it might be wrong but the idea should be correct):
```cpp
void mouse_move_loop( std::function<void (cata::optional<tripoint>)> mouse_pos_cb,
                      std::function<tripoint (void)> edge_scroll_cb )
```
The two `MOUSE_MOVE` loops added in this PR only requires `edge_scroll_cb`, but for the one in `game::look_around()` we need both callbacks, so the loop would be replaced with this:
```cpp
mouse_move_loop(
    []( cata::optional<tripoint> pos ) -> void { if ( !pos) { return; } lx = pos->x; ly = pos->y; },
    []( tripoint scroll ) -> void { center += scroll; },
)
```
Elegant, but certainly requires more cerebral processing once you stumble upon it. Not sure I like it for that reason. What do you think? Any alternatives?

Reply:
> Elegant, but certainly requires more cerebral processing once you stumble upon it. Not sure I like it for that reason. What do you think? Any alternatives?

I wouldn't call this elegant: these lambda arguments overcomplicate things for no good reason. What you want is proper separation of concerns. I'd introduce some function to consume a certain number of specified actions, something like this:
```c++
std::string input_context::consume_actions_n( const std::string &action, int n, int timeout ) {
    std::string action;

    do {
	action = handle_input( timeout );    
    } while( --n > 0 && action == target );
        
    return action;
}
```
And then reuse it:

```c++
const tripoint edge_scroll = g->mouse_edge_scrolling( ctxt, ... );
action = ctxt.consume_actions_n( "MOUSE_MOVE", 10, 10 );
```
Ideally, the scrolling processing function should be implemented as a virtual member of a `window` (or `widget`) class which we don't currently have as such (`catacurses::window` is a pure wrapper).

Reply:
What you propose will work fine for the two cases added in this PR since they only deal with edge scrolling and want them to be rate limited to only happen once every n `MOUSE_MOVE` actions. 

But if we apply your solution in the `"MOUSE_MOVE"` case in `game::look_around()` we get a new behavior: the code updating the mouse cursor cannot be run inside  `input_context::consume_actions_n()`'s loop, so we can only call it once instead of each iteration -- it is also now rate limited. As you can imagine, the result is that the cursor often gets stuck behind the real mouse position.

I believe an abstraction that works for all three cases (look around, target, overmap) will require injecting code into the loop, for instance via lambdas. Unless you have another idea, do you want me to add your solution for only the two ones added in this PR?

Reply:
I noticed that these loops are unnecessary, so this point is now moot.